### ISSUE TITLE
Several operations (simplify, Derivative, pickle) fail for constant Polys

### ISSUE THREAD

**Issue by Adrien79** [sympy/sympy#9918]
Pickle succeeds to store Polys:

```
>>> x = Symbol('x')
>>> P = Poly(x-1)
>>> pickle.loads(pickle.dumps(P))
Poly(x - 1, x, domain='ZZ')
```

But it fails for constant Polys:

```
>>> x = Symbol('x')
>>> Q = Poly(1,x)
>>> pickle.loads(pickle.dumps(Q))
Traceback (most recent call last):
File "<pyshell#11>", line 1, in <module>
  pickle.loads(pickle.dumps(Q))
File "C:\Python34\lib\site-packages\sympy\polys\polytools.py", line 88, in __new__
  return cls._from_expr(rep, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polytools.py", line 198, in _from_expr
  rep, opt = _dict_from_expr(rep, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 332, in _dict_from_expr
  rep, gens = _dict_from_expr_no_gens(expr, opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 277, in _dict_from_expr_no_gens
  (poly,), gens = _parallel_dict_from_expr_no_gens((expr,), opt)
File "C:\Python34\lib\site-packages\sympy\polys\polyutils.py", line 238, in _parallel_dict_from_expr_no_gens
  raise GeneratorsNeeded("specify generators to give %s a meaning" % arg)
sympy.polys.polyerrors.GeneratorsNeeded: specify generators to give 1 a meaning
```

My guess is that the generator is dropped when pickling, and unpickling tries to evaluate `Poly(1)`, an expression which fails with the exact same error message when passed to the shell.

**Comment by Adrien79** [comment]
I've just checked this:

    >>> x = Symbol('x')
    >>> P = Poly(1,x)
    >>> srepr(P)
    Poly(Integer(1))

Then `sympify(srepr(P))` fails with the same error message. So the problem is not specific to Pickle.

**Comment by jksuom** [comment]
There is no method for printing a ``Poly`` in the ``ReprPrinter`` class. Hence it defaults to ``emptyPrinter`` that mistakenly prints only the argument omitting ``gens`` and ``domain``. Apparently it should use ``str(expr)`` instead.

**Comment by Adrien79** [comment]
I've tried this on 0.7.7.dev too.
The weird behaviour of constant Polys is not limited to `srepr` and `pickle`. Other commands fail too, like `simplify(Poly(1,x))` or `Derivative(Poly(1,x), x).doit()` (however `Poly(1,x).diff(x)` does work.)

**Comment by hackalog** [comment]
It looks to me like any generators that are not explicitly referenced in the polynomial are dropped (even on a copy). This means `copy()` returns an unequal Poly,

e.g.:
```
>>> import sympy
>>> from sympy.abc import x, y, z

>>> obj = sympy.Poly('x+y', x, y, z)
>>> o2 = obj.copy()
>>> obj == o2
False
>>> repr(obj), repr(o2)
("Poly(x + y, x, y, z, domain='ZZ')", "Poly(x + y, x, y, domain='ZZ')")
```
The same thing happens after pickling

**Comment by hackalog** [comment]
I've made a potential fix in PR #15799 . Let me know if this fixes your issue

**Comment by czgdp1807** [comment]
Current status,
```python
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(SymPyConsole)
>>> x = Symbol('x')
>>> P = Poly(x-1)
>>> import pickle
>>> pickle.loads(pickle.dumps(P))
Poly(x - 1, x, domain='ZZ')
>>> x = Symbol('x')
>>> Q = Poly(1,x)
>>> pickle.loads(pickle.dumps(Q))
Poly(1, x, domain='ZZ')
>>> x = Symbol('x')
>>> P = Poly(1,x)
>>> srepr(P)
Poly(Integer(1), Symbol('x'))
>>> sympify(srepr(P))
Poly(1, x, domain='ZZ')
```
It looks like the fix has been made. Closing this for now. Please let us know if there is a need to re-open this. Thanks for your contributions.
